### PR TITLE
feat: add Moq1208 analyzer for Returns() delegate mismatch on async methods

### LINKED ISSUE #767: Explainer: Detect Returns() with Non-Task Delegate on Async Methods
## Summary

This explainer proposes a new Roslyn analyzer to detect a common mistake when setting up async methods in Moq: using `.Returns()` with a delegate that returns a non-Task type when the mocked method returns `Task<T>` or `ValueTask<T>`. This pattern causes a runtime exception that can be caught at compile time.

**Problem:** Developers sometimes write `mock.Setup(x => x.AsyncMethod()).Returns(() => value)` when they should use `.ReturnsAsync(() => value)`, leading to confusing runtime errors.

**Solution:** Provide compile-time detection and automatic code fix to guide developers toward the correct pattern.

## Goals

1. **Detect runtime exceptions at compile time** - Catch delegate type mismatches before code runs
2. **Provide clear diagnostic messages** - Help developers understand what went wrong
3. **Offer automatic code fixes** - Transform incorrect patterns to correct ones
4. **Support all async patterns** - Handle `Task<T>`, `ValueTask<T>`, and edge cases
5. **Maintain backward compatibility** - Work correctly across Moq versions (4.8.2+)

## Non-Goals

- Detecting other async setup patterns (already covered by Moq1201, Moq1206)
- Enforcing specific async method naming conventions
- Analyzing non-Moq mocking frameworks

## The Problem

When setting up async methods in Moq, developers sometimes use `.Returns()` with a delegate that returns the unwrapped value instead of a Task. This causes a runtime exception.

**Problem Scenario:**

```csharp
public interface IFoo
{
    Task<int> BarAsync();
}

var m = new Mock<IFoo>(MockBehavior.Strict);
m.Setup(d => d.BarAsync()).Returns(() => { return 7; }); // ❌ RUNTIME ERROR
await m.Object.BarAsync();
```

**Runtime Exception:**

```text
MockException: Invalid callback. Setup on method with return type 'Task' cannot invoke callback with return type 'int'.
```

**Why It Fails:**

- `BarAsync()` returns `Task<int>`
- The delegate `() => 7` returns `int`
- Moq cannot convert `int` to `Task<int>` implicitly
- This is a **type system violation** caught at runtime, not compile time

## User Stories

**As a developer using Moq:**
- I want compile-time warnings when I use Returns() incorrectly with async methods
- So that I can fix the issue before encountering runtime errors

**As a code reviewer:**
- I want the analyzer to catch these mistakes automatically
- So that I don't have to manually check every async setup

**As a team lead:**
- I want consistent patterns across the codebase for async setups
- So that code is more maintainable and less error-prone

## Correct Solutions

**Option 1: Use ReturnsAsync()** (Recommended)

```csharp
m.Setup(d => d.BarAsync()).ReturnsAsync(() => { return 7; });
// or simpler:
m.Setup(d => d.BarAsync()).ReturnsAsync(7);
```

**Option 2: Return Task explicitly** (Verbose)

```csharp
m.Setup(d => d.BarAsync()).Returns(() => Task.FromResult(7));
```

**Option 3: Setup .Result property** (Moq 4.16.0+)

```csharp
m.Setup(d => d.BarAsync().Result).Returns(() => { return 7; });
```

**References:**

- [Moq Quickstart: Async Methods](https://github.com/devlooped/moq/wiki/Quickstart#async-methods)

## Moq API Context

Moq provides these methods for async setups:

- `Setup()` - Standard setup method for all member types (sync and async)
- `ReturnsAsync()` - Extension methods specifically for `Task<T>` and `ValueTask<T>`
- `ThrowsAsync()` - Extension methods for async exceptions
- `Returns()` - Works with any return type, but requires proper Task wrapping for async

**Source References:**
- [Moq Mock\<T\>.Setup() source](https://github.com/devlooped/moq/blob/main/src/Moq/Mock%601.cs#L411-L451)
- [Moq ReturnsAsync extension source](https://github.com/devlooped/moq/blob/main/src/Moq/ReturnsExtensions.cs#L19-L184)

## Related Existing Analyzers

The Moq.Analyzers repository already has analyzers for similar async patterns:

### Moq1201: SetupShouldNotIncludeAsyncResultAnalyzer

**Purpose:** Detects `.Setup(x => x.AsyncMethod().Result)` anti-pattern

**Code Example:**

```csharp
// ❌ Triggers Moq1201 (before Moq 4.16.0)
mock.Setup(c => c.GetAsync().Result).Returns("value");

// ✅ Recommended
mock.Setup(c => c.GetAsync()).ReturnsAsync("value");
```

**Severity:** Error

**Documentation:** [docs/rules/Moq1201.md](../../docs/rules/Moq1201.md)

**Note:** This analyzer skips analysis for Moq 4.16.0+ where the `.Result` pattern became supported.

### Moq1206: ReturnsAsyncShouldBeUsedForAsyncMethodsAnalyzer

**Purpose:** Detects `.Returns(async () => value)` when should use `.ReturnsAsync()`

**Code Example:**

```csharp
// ❌ Triggers Moq1206
mock.Setup(c => c.GetAsync()).Returns(async () => "value");

// ✅ Recommended
mock.Setup(c => c.GetAsync()).ReturnsAsync(() => "value");
```

**Severity:** Warning

**Documentation:** [docs/rules/Moq1206.md](../../docs/rules/Moq1206.md)

### Moq1203: MethodSetupShouldSpecifyReturnValueAnalyzer

**Purpose:** Ensures method setups specify return values (Returns/Throws)

**Implementation:** Uses IOperation-based analysis to detect incomplete setups

**Documentation:** Source code in `src/Analyzers/MethodSetupShouldSpecifyReturnValueAnalyzer.cs`

## Gap Analysis: What's Missing?

Neither Moq1201 nor Moq1206 detects the **specific problem** identified in issue #767:

**The Gap:**

```csharp
// ❌ This pattern is NOT detected by existing analyzers
mock.Setup(x => x.AsyncMethod()).Returns(() => synchronousValue);
//                                       ^^^^^^^^^^^^^^^^^^^^^^^^
//                                       Non-Task returning delegate
```

**Why Current Analyzers Don't Catch It:**

- **Moq1201:** Only looks for `.Result` in the Setup expression
- **Moq1206:** Only looks for `async` lambdas in Returns()
- **Neither checks:** Delegate return type compatibility with method return type

### Analyzer ID: Moq12XX (TBD - assign next available in 1200-1299 range)

**Title:** "ReturnsAsync should be used for async methods instead of Returns with non-Task delegate"

**Category:** Correctness (Moq1200-1299 range)

**Problem Statement:**
Detect when `.Setup()` on an async method uses `.Returns()` with a delegate that returns a non-Task type, which will cause a runtime exception.

**Detection Pattern:**

1. **Identify Setup invocation** on a member returning `Task`, `Task<T>`, `ValueTask`, or `ValueTask<T>`
2. **Check if followed by `.Returns()`** invocation
3. **Inspect the delegate signature** passed to Returns()
4. **If delegate returns non-Task type:** Report diagnostic

**Diagnostic Message:**

```text
Moq12XX: Async method setup should use ReturnsAsync instead of Returns with non-Task delegate.
Method '{methodName}' returns '{returnType}', but Returns() delegate returns '{delegateReturnType}'.
Use ReturnsAsync() or Returns(() => Task.FromResult(...)) instead.
```

**Severity:** Error (runtime exception guaranteed)

**Code Fix:**

Transform to ReturnsAsync pattern:

```csharp
// Before:
mock.Setup(x => x.GetAsync()).Returns(() => 7);

// After:
mock.Setup(x => x.GetAsync()).ReturnsAsync(() => 7);
```

#### Approach 1: Type-based detection

```csharp
// When incoming type is IReturns<T1, Task<T2>>
// and .Returns() binds to delegate returning non-Task type
// suggest ReturnsAsync
```

#### Approach 2: Delegate signature inspection

```csharp
// If bound type in .Returns() is Delegate
// inspect its return type
// compare against expected Task/ValueTask type
// report mismatch
```

**Recommended:** Combination of both approaches using IOperation API for semantic analysis

### Technical Implementation Details

**Analysis Method:** IOperation-based (like Moq1203)

**Key Steps:**
1. Register `OperationKind.Invocation` action
2. Identify `Setup()` invocations via `MoqKnownSymbols`
3. Extract mocked member's return type from Setup expression
4. Check if return type is `Task`, `Task<T>`, `ValueTask`, or `ValueTask<T>`
5. Find chained `.Returns()` invocation
6. Inspect Returns() argument (delegate expression)
7. Compare delegate return type with expected Task type
8. Report diagnostic on character-precise span if mismatch

**Code Fix Strategy:**
- Replace `Returns` with `ReturnsAsync` in invocation
- Preserve delegate body exactly
- Extract inner value from Task wrapper if present
- Use existing code fix patterns from Moq1206 as template

### Test Data Strategy

**Required Test Groups:**

1. **Moq 4.18.4+ (Current)**
   - `Task<T>` return types
   - `ValueTask<T>` return types
   - Delegate returning T when method returns `Task<T>`
   - Delegate returning `Task<T>` when method returns `Task<T>` (no diagnostic)
   - Async lambda (handled by Moq1206, should not trigger new analyzer)

2. **Moq 4.8.2 (Legacy Compatibility)**
   - `Task<T>` return types only (ValueTask not supported)
   - Same delegate mismatch scenarios
   - Verify no false positives

3. **Edge Cases:**
   - Nested Task types: `Task<Task<T>>`
   - Custom awaitable types (should not trigger)
   - Callback overloads vs Returns overloads
   - Generic methods with Task constraints

**Test Data Pattern** (from existing analyzers):

```csharp
public static IEnumerable<object[]> DetectsReturnsDelegateMismatchData()
{
    return new object[][]
    {
        [
            """public Task<int> GetAsync() => Task.FromResult(0);""",
            """mock.Setup(x => x.GetAsync()).Returns(() => 42);""", // broken
            """mock.Setup(x => x.GetAsync()).ReturnsAsync(() => 42);""", // fixed
        ],
    }.WithNamespaces().WithMoqReferenceAssemblyGroups();
}
```

## Moq Version Compatibility Matrix

| Feature | Moq 4.8.2 | Moq 4.16.0+ | Moq 4.18.4+ |
|---------|-----------|-------------|-------------|
| ReturnsAsync for `Task<T>` | ✅ | ✅ | ✅ |
| ReturnsAsync for `ValueTask<T>` | ❌ | ✅ | ✅ |
| Setup().Result pattern | ❌ | ✅ | ✅ |
| Protected().Setup() | ❌ | ❌ | ✅ |

**Analyzer Behavior:**
- For Moq 4.16.0+: Can suggest `.Setup().Result.Returns()` as alternative
- For Moq 4.18.4+: Include ValueTask scenarios
- For all versions: ReturnsAsync is valid and preferred

### GitHub Issues

- **[#767](https://github.com/rjmurillo/moq.analyzers/issues/767)** - Current issue (incorrect premise, but identifies real problem)
- **[#432](https://github.com/rjmurillo/moq.analyzers/issues/432)** - Marked as duplicate of #767

### Moq Documentation

- [Quickstart: Async Methods](https://github.com/devlooped/moq/wiki/Quickstart#async-methods) - Official guide
- [ReturnsExtensions source](https://github.com/devlooped/moq/blob/main/src/Moq/ReturnsExtensions.cs) - Implementation details

### Repository Documentation

- [Moq1201 Rule](https://github.com/rjmurillo/moq.analyzers/blob/main/docs/rules/Moq1201.md) - Related analyzer (Setup().Result pattern)
- [Moq1206 Rule](https://github.com/rjmurillo/moq.analyzers/blob/main/docs/rules/Moq1206.md) - Related analyzer (async lambda pattern)
- [Moq1200-1299 Category](https://github.com/rjmurillo/moq.analyzers/blob/main/docs/rules/README.md) - Correctness analyzers for async patterns

### Phase 1: Analyzer Core (8-12 hours)

1. **Assign Moq12XX ID** - Next available in 1200-1299 correctness range
2. **Use Moq1203 as template** - IOperation-based analysis approach
3. **Implement version-aware logic** - Handle Moq 4.8.2, 4.16.0+, 4.18.4+ differences
4. **Follow data-driven test pattern** - Use `.WithMoqReferenceAssemblyGroups()`
5. **Add comprehensive documentation** - Create `docs/rules/Moq12XX.md`
6. **Update analyzer releases** - Add entry to `AnalyzerReleases.Unshipped.md`

### Phase 2: Code Fix (4-6 hours)

1. **Transform Returns to ReturnsAsync** - Preserve delegate body
2. **Handle all overloads** - Multiple Returns() signatures
3. **Preserve code formatting** - Character-precise spans
4. **Support batch fixing** - FixAll provider

### Phase 3: Documentation and Testing (4-6 hours)

1. **Create rule documentation** - `docs/rules/Moq12XX.md`
2. **Update analyzer releases** - Add entry to `AnalyzerReleases.Unshipped.md`
3. **Update README** - Add to analyzer list
4. **Performance benchmarks** - Ensure no regression

### Priority and Scope

**Priority:** Medium-High
- Runtime exception (severe)
- Common mistake
- Clear fix available

**Estimated Effort:** 16-24 hours
- 8-12 hours: Analyzer implementation and unit tests
- 4-6 hours: Code fix implementation
- 2-4 hours: Documentation and integration
- 2-4 hours: Performance benchmarks and validation

**Complexity:** Medium
- IOperation-based analysis (similar to Moq1203)
- Delegate signature inspection (new pattern)
- Version-aware behavior (established pattern)

## Success Metrics

**Detection Accuracy:**
- 95%+ true positive rate on real-world codebases
- Less than 1% false positive rate
- No false negatives on tested patterns

**Code Fix Quality:**
- 100% of fixes produce compilable code
- Preserve original code formatting and style
- Handle all Returns() overloads correctly

**Performance:**
- No measurable impact on build time for typical projects
- Analysis completes within 100ms for single file

## Open Questions

1. **Should we detect Returns(() => Task.FromResult(value))?**
   - This pattern is verbose but valid
   - Consider info-level suggestion to simplify to ReturnsAsync?

2. **How to handle complex delegate expressions?**
   - Delegate with multiple statements
   - Delegate with async/await inside (covered by Moq1206)

3. **Edge case: `Task<Task<T>>` scenarios?**
   - Rare but possible
   - May need special handling or explicit exclusion

## Conclusion

This analyzer addresses a common pain point when using Moq with async methods. By detecting type mismatches at compile time, it prevents runtime exceptions and guides developers toward best practices.

The implementation leverages existing patterns from Moq1203 (IOperation-based analysis) and Moq1206 (async pattern detection), making it a natural addition to the Moq1200-1299 correctness analyzer category.

### PR DESCRIPTION
**Pull request by rjmurillo**
## Summary

- Adds new analyzer **Moq1208** that detects `.Returns(() => value)` with a delegate whose return type does not match the async method's return type (e.g., returns `int` instead of `Task<int>`), which causes a runtime `MockException`
- Adds code fix to transform `.Returns(delegate)` to `.ReturnsAsync(delegate)`
- Includes 100 tests (82 analyzer + 18 code fix) covering all branches, edge cases, and doppelganger scenarios

## Problem

```csharp
mock.Setup(d => d.BarAsync()).Returns(() => 7); // Runtime MockException!
// "Invalid callback. Setup on method with return type 'Task' cannot invoke callback with return type 'int'."
```

## Solution

The analyzer flags the mismatch at compile time. The code fix transforms:
```csharp
// Before (flagged)
mock.Setup(d => d.BarAsync()).Returns(() => 7);

// After (fixed)
mock.Setup(d => d.BarAsync()).ReturnsAsync(() => 7);
```

## Files changed

| File | Change |
|------|--------|
| `src/Common/DiagnosticIds.cs` | Added `ReturnsDelegateMismatchOnAsyncMethod = "Moq1208"` |
| `src/Analyzers/ReturnsDelegateShouldReturnTaskAnalyzer.cs` | New analyzer |
| `src/CodeFixes/ReturnsDelegateShouldReturnTaskFixer.cs` | New code fix |
| `src/Analyzers/AnalyzerReleases.Unshipped.md` | Release tracking |
| `tests/.../ReturnsDelegateShouldReturnTaskAnalyzerTests.cs` | 82 test cases |
| `tests/.../ReturnsDelegateShouldReturnTaskFixerTests.cs` | 20 test cases (5 x 2 namespaces x 2 Moq versions) |
| `docs/rules/Moq1208.md` | Rule documentation |

## Edge cases handled

| Case | Behavior |
|------|----------|
| `.Returns(value)` (direct value, not delegate) | Not flagged |
| `.Returns(() => Task.FromResult(7))` | Not flagged (correct) |
| `.Returns(async () => 7)` | Not flagged (Moq1206 handles this) |
| Non-async method `.Returns(() => 7)` | Not flagged |
| Non-generic `Task` return type | Not flagged |
| `.Callback().Returns(() => 7)` on async method | Flagged |
| `(mock.Setup(...)).Returns(() => 7)` | Flagged |
| Delegate with parameters `(string x) => x.Length` | Flagged |

## Test plan

- [x] Build succeeds with 0 warnings, 0 errors
- [x] 100 new Moq1208 tests pass
- [x] Full test suite passes (2604/2604, 0 failures, 0 regressions)
- [x] Doppelganger tests verify no false positives on custom Mock classes
- [x] Tests cover all analyzer branches for 100% block coverage
- [x] Tests run across multiple Moq versions (WithMoqReferenceAssemblyGroups)

Closes #767

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added analyzer Moq1208 (warning) detecting Returns() delegates that mismatch async setups and a quick-fix to replace with ReturnsAsync().
* **Documentation**
  * Added rule documentation and updated rules index with examples, recommended fixes, and suppression/configuration guidance.
* **Tests**
  * Added comprehensive analyzer and code-fix tests covering valid/invalid patterns for Task/ValueTask and delegate forms.
* **Chores**
  * Added release-note entry and minor test-verifier refactor to support optional compiler diagnostics.

### CODE REVIEW

File: src/Analyzers/ReturnsDelegateShouldReturnTaskAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        // Get the type info of the lambda's body expression to determine what type it returns
+        ExpressionSyntax? bodyExpression = lambda.Body as ExpressionSyntax;
+        if (bodyExpression == null)
+        {
+            return null;
+        }
+
+        TypeInfo typeInfo = semanticModel.GetTypeInfo(bodyExpression);
+        return typeInfo.Type;
+    }
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply src/Analyzers/ReturnsDelegateShouldReturnTaskAnalyzer.cs]: Adopted this approach. `GetDelegateReturnType` now uses `semanticModel.GetSymbolInfo(lambda).Symbol as IMethodSymbol` to get `ReturnType`, which handles expression-bodied, block-bodied, and multi-return lambdas uniformly. Also removed `FindFirstReturnStatement` since it is no longer needed. Fixed in 6a07ef9.

File: src/Analyzers/ReturnsDelegateShouldReturnTaskAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+        // Must be a lambda or delegate. Raw values (not delegates) are a different overload.
+        if (firstArgument is not LambdaExpressionSyntax lambda)
+        {
+            return false;
+        }
+
+        // Exclude async lambdas. Those are Moq1206's domain.
+        return !lambda.AsyncKeyword.IsKind(SyntaxKind.AsyncKeyword);
+    }
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply src/Analyzers/ReturnsDelegateShouldReturnTaskAnalyzer.cs]: Valid observation. Method groups and `AnonymousMethodExpressionSyntax` are not detected. This is a deliberate scope limitation for the initial implementation since lambda expressions are by far the most common pattern. Tracking as a follow-up enhancement.

File: docs/rules/Moq1208.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
+# Moq1208: Returns() delegate type mismatch on async method setup
+
+| Item     | Value |
+| -------- | ----- |
+| Enabled  | True  |
+| Severity | Warning |
+| CodeFix  | True |
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply docs/rules/Moq1208.md]: The table uses standard single pipes matching the format in Moq1206.md and other peer docs. This comment may be based on an earlier revision. The current Moq1208.md was rewritten in bd782b2 and uses correct formatting.

File: src/Analyzers/ReturnsDelegateShouldReturnTaskAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+/// <summary>
+/// Returns() delegate on async method setup should return Task/ValueTask to match the mocked method's return type.
+/// </summary>
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public class ReturnsDelegateShouldReturnTaskAnalyzer : DiagnosticAnalyzer
+{
+    private static readonly LocalizableString Title = "Moq: Returns() delegate type mismatch on async method";
+    private static readonly LocalizableString Message = "Returns() delegate for async method '{0}' should return a Task type, not '{1}'. Use ReturnsAsync() or return Task.FromResult().";
+    private static readonly LocalizableString Description = "Returns() delegate on async method setup should return Task/ValueTask. Use ReturnsAsync() or wrap with Task.FromResult().";
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply src/Analyzers/ReturnsDelegateShouldReturnTaskAnalyzer.cs]: Agreed. The diagnostic message now includes the expected return type as a third placeholder: `"Returns() delegate for async method '{0}' should return '{2}', not '{1}'. Use ReturnsAsync() or wrap with Task.FromResult()."`. Fixed in 6a07ef9.

File: src/Analyzers/ReturnsDelegateShouldReturnTaskAnalyzer.cs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+            current = candidateMemberAccess.Expression;
+        }
+    }
+
+    private static bool TryGetMismatchInfo(
+        InvocationExpressionSyntax setupInvocation,
+        InvocationExpressionSyntax returnsInvocation,
+        SemanticModel semanticModel,
+        MoqKnownSymbols knownSymbols,
+        [NotNullWhen(true)] out string? methodName,
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply src/Analyzers/ReturnsDelegateShouldReturnTaskAnalyzer.cs]: Acknowledged. The parameter count increased by one (added `expectedReturnType` output) to support the improved diagnostic message. The out-parameter pattern is standard for Roslyn `TryGet` methods. Refactoring to reduce parameters is tracked for follow-up.

File: src/Analyzers/ReturnsDelegateShouldReturnTaskAnalyzer.cs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        return null;
+    }
+
+    private static bool TryGetMismatchInfo(
+        InvocationExpressionSyntax setupInvocation,
+        InvocationExpressionSyntax returnsInvocation,
+        SemanticModel semanticModel,
+        MoqKnownSymbols knownSymbols,
+        [NotNullWhen(true)] out string? methodName,
+        [NotNullWhen(true)] out ITypeSymbol? expectedReturnType,
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply src/Analyzers/ReturnsDelegateShouldReturnTaskAnalyzer.cs]: Acknowledged. The parameter count (7) includes 4 inputs and 3 out-parameters, which is standard for Roslyn TryGet methods that extract multiple pieces of information atomically.

File: src/Analyzers/ReturnsDelegateShouldReturnTaskAnalyzer.cs [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+    /// <inheritdoc />
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+
+    /// <inheritdoc />
+    public override void Initialize(AnalysisContext context)
+    {
+        context.EnableConcurrentExecution();
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.RegisterSyntaxNodeAction(Analyze, SyntaxKind.InvocationExpression);
+    }
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply src/Analyzers/ReturnsDelegateShouldReturnTaskAnalyzer.cs]: Valid architectural direction. Migrating to `IOperation` would reduce manual `GetSymbolInfo` calls. However, this is an enhancement, not a correctness issue. The current `RegisterSyntaxNodeAction` approach is consistent with peer analyzers (Moq1200, Moq1206). Tracking as a follow-up refactor.
